internal/shim-sgx/src/heap.rs

```diff
@@ -96,35 +97,80 @@ impl<'a> Heap<'a> {
             .unwrap_or(self.memory().as_ptr() as _)
     }
 
+    fn pos_max(&self) -> usize {
+        self.brk_max
+            .map(|x| x.into())
+            .unwrap_or(self.memory().as_ptr() as _)
+    }
+
+    #[cfg(not(test))]
+    fn mmap_augment(&mut self, addr: c_size_t, length: c_size_t) -> Result<usize, c_int> {
+        use sgx::page::{Class, Flags};
+        use x86_64::addr::VirtAddr;
+        use x86_64::structures::paging::Page as PageAddr;
+
+        let pages = (length + Page::SIZE - 1) / Page::SIZE;
+        let secinfo = Class::Regular.info(Flags::READ | Flags::WRITE | Flags::EXECUTE);
+        let zero = PageAddr::from_start_address(VirtAddr::new(ZERO_PAGE.as_ptr() as u64)).unwrap();
+        for i in 0..pages {
+            let paddr = PageAddr::from_start_address(VirtAddr::new((addr + i * Page::SIZE) as u64))
+                .unwrap();
+            if secinfo.accept_copy(paddr, zero).is_err() {
+                panic!();
+            }
+        }
+        Ok(addr)
+    }
+
+    fn mmap_fixed(
+        &mut self,
+        addr: Address<usize, Page>,
+        length: Offset<usize, Page>,
+        access: Access,
+        ledger: usize,
+    ) -> Result<usize, c_int> {
+        let start = self.memory().as_ptr() as usize;
+        let end = start + self.memory.len() * Page::SIZE;
+
+        assert!(length.items() > 0);
+        assert!(addr.raw() >= start);
+        assert!((addr.raw() + length.items() * Page::SIZE) <= end);
+
+        let region = Region::new(
+            Address::new(addr.raw() - start),
+            Address::new(addr.raw() - start) + length,
+        );
+
+        if self.ledgers[ledger].map(region, access).is_err() {
+            return Err(EACCES);
+        }
+
+        #[cfg(not(test))]
+        if ledger == 0 {
+            return self.mmap_augment(addr.raw(), length.bytes());
+        }
```

Comment:
why is this cfg? does this affect integration tests?

Reply:
Because of two tests inside `heap.rs`. All test flagging is because of that. I don't know if that affect integration tests or not. I assumed that not. We have so good coverage (with`rstest`) in `mmledger` already so maybe we could just delete those tests?

Reply:
Let's see what happens.

Reply:
https://github.com/enarx/enarx/pull/1608/commits/5d47df9e4e2f6c8a9c1ba93292226b658b3a44e3

Reply:
Outputting address, in addition to the interrupt vector IMHO is essential. Really helped combined with the (post-processed) log file.

Reply:
a test: https://github.com/enarx/enarx/pull/1608/commits/68874a81a87cc09efc8900e11561e64e9d2d6b86